hera_cal/frf.py

```diff
@@ -575,12 +890,12 @@ def filter_data(self, data, frps, flags=None, nsamples=None,
             filt_flags[k] = f
             filt_nsamples[k] = eff_nsamples
 
-    def tophat_frfilter(self, keys=None, wgts=None, mode='clean',
+    def tophat_frfilter(self, keys=None, wgts=None, mode='clean', uvb=None, percentile_low=5., percentile_high=95.,
```

Comment:
I feel like this function has two _very_ different operating modes (with and w/o the beam). It strikes me that you should write the version that uses the beam as a separate function that calls this one.

Reply:
I think I agree with Josh on this.

Reply:
I still think 3-4 functions might be cleaner than 1 (and also more modular, since you could calculate the fringe rate limits without actually having to do the filter). That said, if you really want to leave it as one function, perhaps you can group the kwargs into sections for different operating modes? It's really hard to tell which options are relevant in which cases and which options are necessary when others are/aren't provided. 

Also, I don't get the line "if provided, will supercede main-lobe" What does that mean?

Finally, can you make the tabbing of the docstring consistent? It's very hard to read right now.

Reply:
perhaps I'm echoing the sentiment here: we have methods for applying an FIR filter given a fringe-rate profile (FRP) in frf.py already. currently, if you provide kwargs (e.g. frate_standoff, max_frate_coeffs) then it constructs the FRP and uses the existing code to apply the filter. part of the new additions is to generate an FRP given a beam, but I think we will likely want to filter with any arbitrary FRP in the near future, not just the ones generated by the function here. I think this is possibly simpler if you just have this high-level function take existing arguments *or* an `frp` dictionary, which is user provided and supercedes the existing arguments. Your existing build_frate() function is a nice interface between `uvb` beam objects and a dictionary of `frp` profiles, so this connection can be made in the run script. I think this would help solve the issue of different operating modes, b/c now there are only two: give me an frp you've made, or have me make a simple one given your choice of kwargs.

Obviously we've got four people here with four slightly differing opinions, but I do think that maintaining this function as-written may be difficult for people other than AEW. lmk your thoughts aaron, thanks!

Reply:
Other than this thread, I'm fine with this PR. Happy to find a time to chat this afternoon to figure out a resolution to makes the three of us happy, @aewallwi @nkern

Reply:
> perhaps I'm echoing the sentiment here: we have methods for applying an FIR filter given a fringe-rate profile (FRP) in frf.py already. currently, if you provide kwargs (e.g. frate_standoff, max_frate_coeffs) then it constructs the FRP and uses the existing code to apply the filter. part of the new additions is to generate an FRP given a beam, but I think we will likely want to filter with any arbitrary FRP in the near future, not just the ones generated by the function here. I think this is possibly simpler if you just have this high-level function take existing arguments _or_ an `frp` dictionary, which is user provided and supercedes the existing arguments. Your existing build_frate() function is a nice interface between `uvb` beam objects and a dictionary of `frp` profiles, so this connection can be made in the run script. I think this would help solve the issue of different operating modes, b/c now there are only two: give me an frp you've made, or have me make a simple one given your choice of kwargs.

@nkern I think I've addressed your concern here by providing an `frate_profiles` arg to `tophat_frfilter`. Let me know if this is good. 

Perhaps I'm not quite parsing what people are requesting here. Do people want this function to only take in frate_centers and hfrate_half_widths dicts (technically the core of what is required for rectangular fringe-rate filtering) and then have a helper function that determines these from the different possible options (uvbeams, max_frate_coeffs, frate_standoff etc..) and then have a more general fringe-rate filtering interface function that supports the tophat mode but also has hooks to support applying multiplicative (in fringe-rate) fir filters?

Reply:
Yeah, this function just has a bunch of different functionalities in its kwargs its kind of hard to use. Perhaps as you suggest it would be easier for this function to just take frate_centers and frate_half_widths and other kwargs.

Then we have a separate function that maps the variety of kwargs here to the frate_centers and frate_half_widths, which would be a massive multi-pronged function but at least we are upfront that this function is complex, and are making the end-user interface (tophat_frfilter) a little cleaner. effectively, revert this function almost back to its state on master (except for a few added features like center_frates) and push all the new guts to a new "usecase_determination_function". here is an example of what I mean

```
def tophat_frfilter(self, keys=None, ..., frate_centers=None, frate_half_widths=None, ...., **filter_kwargs):
    ...


def select_tophat_frates(case, ...):
    """
    Select fringe-rates to filter given a filtering case

    Parameters
    ------------
    case : str
        Filtering case ['frate_coeffs', 'uvbeam', 'fr_profiles']
    """
    if case == 'frate_coeffs':
        if max_frate_coeffs is None:
            center_frates, width_frates = sky_frates(uvd=self.hd, keys=keys, frate_standoff=frate_standoff,
                                                     frate_width_multiplier=frate_width_multiplier, min_frate_half_width=min_frate_half_width)
        else:
            width_frates = {k: np.max([max_frate_coeffs[0] * self.blvecs[k[:2]][0] + max_frate_coeffs[1], 0.0]) for k in keys}
            center_frates = {k: 0.0 for k in keys}

    elif case == 'uvbeam':
        ...

    elif case == 'fr_profiles':
        ...

    else:
        raise ValueError("did not recognize {}".format(case))
    
    return fc, fhw

# for xtalk filter mode as we've done previously we'd do
fc, fhw = select_tophat_frates('frate_coeffs', max_frate_coeffs=[0.01, 0.2], ...)

# for UVBeam inspired, we'd do
fc, fhw = select_tophat_frates('uvbeam', uvb=uvb, percentile_low=0.05, percentile_hi=0.95)

# etc..
```

lmk your thoughts